internal/impl/snowflake/output_snowflake_streaming.go

```diff
@@ -392,135 +501,210 @@ func newSnowflakeStreamer(
-	return o.client.OpenChannel(ctx, streaming.ChannelOptions{
-		ID:                      id,
-		Name:                    name,
-		DatabaseName:            o.db,
-		SchemaName:              o.schema,
-		TableName:               o.table,
-		BuildOptions:            o.buildOpts,
-		StrictSchemaEnforcement: o.schemaEvolutionEnabled(),
-	})
+// SnowflakeClientResourceForTesting is a key that can be used to access the REST client for the snowflake output
+// which can remove boilerplate from tests to setup a new REST client.
+const SnowflakeClientResourceForTesting snowflakeClientForTesting = "SnowflakeClientResourceForTesting"
+
+type dynamicSnowpipeStreamingOutput struct {
+	table   *service.InterpolatedString
+	byTable pool.Indexed[service.BatchOutput]
+
+	initStatementsFn func(context.Context, *streaming.SnowflakeRestClient) error
+	client           *streaming.SnowflakeServiceClient
+	restClient       *streaming.SnowflakeRestClient
 }
 
-func (o *snowflakeStreamerOutput) Connect(ctx context.Context) error {
+func (o *dynamicSnowpipeStreamingOutput) Connect(ctx context.Context) error {
 	if o.initStatementsFn != nil {
 		if err := o.initStatementsFn(ctx, o.restClient); err != nil {
 			return fmt.Errorf("unable to run initialization statement: %w", err)
 		}
 		// We've already executed our init statement, we don't need to do that anymore
 		o.initStatementsFn = nil
 	}
-	// Precreate a single channel so we know stuff works, otherwise we'll create them on demand.
-	c, err := o.channelPool.Acquire(ctx)
-	if err == nil {
-		// We succeeded! Put the channel back and move on.
-		o.channelPool.Release(c)
-		return nil
+	return nil
+}
+
+func (o *dynamicSnowpipeStreamingOutput) WriteBatch(ctx context.Context, batch service.MessageBatch) error {
+	executor := batch.InterpolationExecutor(o.table)
+	tableBatches := map[string]service.MessageBatch{}
+	for i, msg := range batch {
+		table, err := executor.TryString(i)
+		if err != nil {
+			return fmt.Errorf("unable to interpolate `%s`: %w", ssoFieldTable, err)
+		}
+		tableBatches[table] = append(tableBatches[table], msg)
 	}
-	// It's possible we couldn't open the channel because the table doesn't exist - let's check that.
-	autoCreate, autoCreateErr := o.WillAutoCreateTable(ctx)
-	// If we know we don't have an output table and schema evolution is enabled, then don't create
-	// any channels - we'll do it lazily after we can create the table (which we need data for).
-	if autoCreateErr == nil && autoCreate {
-		o.logger.Debug("determined table does not exist, waiting to auto-create the table until we have data")
-		return nil
+	for table, batch := range tableBatches {
+		output, err := o.byTable.Acquire(ctx, table)
+		if err != nil {
+			return err
+		}
+		// Immediately release, these are thread safe, so we can let other
+		// threads modify them while we have a reference.
+		o.byTable.Release(table, output)
+		if err := output.WriteBatch(ctx, batch); err != nil {
+			return err
+		}
```

Comment:
Nit: Might be handy to annotate these errors

Reply:
These are already annotated a layer down :)